File: leetcode/0325-maximum-size-subarray-sum-equals-k/solution.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxSubArrayLen(vector<int>& nums, int k) {
        unordered_map<long long, int> firstSeen;
        firstSeen[0] = -1;
        long long prefix = 0;
        int best = 0;
        for (int i = 0; i < (int)nums.size(); i++) {
            prefix += nums[i];
            long long needed = prefix - k;
            auto it = firstSeen.find(needed);
            if (it != firstSeen.end()) {
                best = max(best, i - it->second);
            }
            if (firstSeen.find(prefix) == firstSeen.end()) {
                firstSeen[prefix] = i;
            }
        }
        return best;
    }
};

File: leetcode/0325-maximum-size-subarray-sum-equals-k/solution.py
from typing import List


class Solution:
    def maxSubArrayLen(self, nums: List[int], k: int) -> int:
        first_seen = {0: -1}
        prefix = 0
        best = 0
        for i, x in enumerate(nums):
            prefix += x
            needed = prefix - k
            if needed in first_seen:
                best = max(best, i - first_seen[needed])
            if prefix not in first_seen:
                first_seen[prefix] = i
        return best
